File: python/problems/leetcode/0008_string_to_integer_atoi/solution.py
"""LeetCode 8: String to Integer (atoi)."""

from __future__ import annotations

INT_MIN, INT_MAX = -(2**31), 2**31 - 1


class Solution:
    def myAtoi(self, s: str) -> int:
        i, n = 0, len(s)
        while i < n and s[i] == " ":
            i += 1
        if i == n:
            return 0
        sign = 1
        if s[i] in "+-":
            if s[i] == "-":
                sign = -1
            i += 1
        x = 0
        while i < n and s[i].isdigit():
            x = x * 10 + ord(s[i]) - 48
            if sign * x > INT_MAX:
                return INT_MAX
            if sign * x < INT_MIN:
                return INT_MIN
            i += 1
        return sign * x


if __name__ == "__main__":
    assert Solution().myAtoi("42") == 42
    assert Solution().myAtoi("   -42") == -42
    assert Solution().myAtoi("4193 with words") == 4193
    assert Solution().myAtoi("words and 987") == 0
    assert Solution().myAtoi("-91283472332") == INT_MIN
    assert Solution().myAtoi("+1") == 1
    print("0008 OK")

File: cpp/problems/leetcode/0008_string_to_integer_atoi/solution.cpp
// LeetCode 8
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

class Solution {
public:
    int myAtoi(string s) {
        int i = 0, n = (int)s.size();
        while (i < n && s[i] == ' ') ++i;
        if (i == n) return 0;
        int sign = 1;
        if (s[i] == '+' || s[i] == '-') {
            if (s[i] == '-') sign = -1;
            ++i;
        }
        long long x = 0;
        while (i < n && isdigit(static_cast<unsigned char>(s[i]))) {
            x = x * 10 + (s[i] - '0');
            if (sign * x > INT_MAX) return INT_MAX;
            if (sign * x < INT_MIN) return INT_MIN;
            ++i;
        }
        return static_cast<int>(sign * x);
    }
};

int main() {
    assert(Solution().myAtoi("42") == 42);
    assert(Solution().myAtoi("   -42") == -42);
    assert(Solution().myAtoi("4193 with words") == 4193);
    assert(Solution().myAtoi("words and 987") == 0);
    assert(Solution().myAtoi("-91283472332") == INT_MIN);
    assert(Solution().myAtoi("+1") == 1);
    cout << "0008 OK" << endl;
    return 0;
}
